ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoiceNotifier.java

```diff
@@ -0,0 +1,289 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.statetransition.forkchoice;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.spec.Spec;
+import tech.pegasys.teku.spec.datastructures.blocks.StateAndBlockSummary;
+import tech.pegasys.teku.spec.datastructures.operations.versions.merge.BeaconPreparableProposer;
+import tech.pegasys.teku.spec.datastructures.state.beaconstate.BeaconState;
+import tech.pegasys.teku.spec.executionengine.ExecutionEngineChannel;
+import tech.pegasys.teku.spec.executionengine.ForkChoiceState;
+import tech.pegasys.teku.spec.executionengine.ForkChoiceUpdatedResult;
+import tech.pegasys.teku.spec.executionengine.PayloadAttributes;
+import tech.pegasys.teku.ssz.type.Bytes20;
+import tech.pegasys.teku.storage.client.RecentChainData;
+
+public class ForkChoiceNotifier {
+  private static final int MAX_PROPOSER_SEEN_EPOCHS = 2;
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final ExecutionEngineChannel executionEngineChannel;
+  private final Spec spec;
+  private final RecentChainData recentChainData;
+
+  private final CurrentForkChoiceUpdatedInfo currentForkChoiceUpdatedInfo =
+      new CurrentForkChoiceUpdatedInfo();
+
+  private Map<UInt64, ProposerLastSeenSlotAndFeeRecipient>
+      proposerIndexLastSeenSlotAndFeeRecipient = new HashMap<>();
+
+  private Optional<UInt64> maybeLastAttestationDueSlot = Optional.empty();
+
+  public ForkChoiceNotifier(
+      final RecentChainData recentChainData,
+      final ExecutionEngineChannel executionEngineChannel,
+      final Spec spec) {
+    this.recentChainData = recentChainData;
+    this.executionEngineChannel = executionEngineChannel;
+    this.spec = spec;
+  }
+
+  public void onUpdatePreparableProposers(
+      Collection<BeaconPreparableProposer> beaconPreparableProposers) {
+    LOG.debug("onUpdatePreparableProposers {}", beaconPreparableProposers);
+
+    recentChainData
+        .getCurrentSlot()
+        .ifPresent(
+            currentSlot ->
+                updatePreparableProposers(
+                    beaconPreparableProposers,
+                    attestationDueHasBeenCalledForSlot(currentSlot)
+                        ? currentSlot.plus(1)
+                        : currentSlot));
```

Comment:
to recap the ideas around this:
 - since this is called at the beginning of the slot, we still may have time to propose for the current slot.  (but requires _at least_ that we move `BeaconProposerPreparer` before `BlockProductionDutyLoader` in the `validatorTimingChannels` list). the slot+1 will be in any case guaranteed by `onForkChoiceUpdated` or the upcoming `onAttestationsDue`

 - to relax assumptions on when the method is allowed to be called, I also cover the case when is called after `onAttestationsDue`, that is too late for current slot and we deal with the next.

but if the first case is unrealistic, everything will fall under the +1 case